wpilibc/src/main/native/include/frc/LEDPattern.h

```diff
@@ -0,0 +1,325 @@
+   * that illuminates only the portion of the LED strip corresponding with some
+   * progress. The mask pattern will start from the base and set LEDs to white
+   * at a proportion equal to the progress returned by the function. Some usages
+   * for this could be for displaying progress of a flywheel to its target
+   * velocity, progress of a complex autonomous sequence, or the height of an
+   * elevator.
+   *
+   * <p>For example, creating a mask for displaying a red-to-blue gradient,
+   * starting from the red end, based on where an elevator is in its range of
+   * travel.
+   *
+   * <pre>
+   *   frc::LEDPattern basePattern = frc::LEDPattern::Gradient(frc::Color::kRed,
+   * frc::Color::kBlue); frc::LEDPattern progressPattern =
+   *     basePattern.Mask(frc::LEDPattern::ProgressMaskLayer([&]() { return
+   * elevator.GetHeight() / elevator.MaxHeight(); });
+   * </pre>
+   *
+   * @param progressFunction the function to call to determine the progress.
+   * This should return values in the range [0, 1]; any values outside that
+   * range will be clamped.
+   * @return the mask pattern
+   */
+  static LEDPattern ProgressMaskLayer(std::function<double()> progressFunction);
+
+  /**
+   * Display a set of colors in steps across the length of the LED strip. No
+   * interpolation is done between colors. Colors are specified by the first LED
+   * on the strip to show that color. The last color in the map will be
+   * displayed all the way to the end of the strip. LEDs positioned before the
+   * first specified step will be turned off (you can think of this as if
+   * there's a 0 -> black step by default).
+   *
+   * @param steps a map of progress to the color to start displaying at that
+   * position along the LED strip
+   * @return a motionless step pattern
+   */
+  static LEDPattern Steps(std::span<std::pair<double, Color>> steps);
+
+  /**
+   * Creates a pattern that displays a non-animated gradient of colors across
+   * the entire length of the LED strip. The gradient wraps around so the start
+   * and end of the strip are the same color, which allows the gradient to be
+   * modified with a scrolling effect with no discontinuities. Colors are evenly
+   * distributed along the full length of the LED strip.
+   *
+   * @param colors the colors to display in the gradient
+   * @return a motionless gradient pattern
+   */
+  static LEDPattern Gradient(std::span<Color> colors);
+
+  /**
+   * Creates an LED pattern that displays a rainbow across the color wheel. The
+   * rainbow pattern will stretch across the entire length of the LED strip.
+   *
+   * @param saturation the saturation of the HSV colors, in [0, 255]
+   * @param value the value of the HSV colors, in [0, 255]
+   * @return the rainbow pattern
+   */
+  static LEDPattern Rainbow(int saturation, int value);
+
+  // floorDiv algorithm taken from Java
+  constexpr static int floorDiv(int x, int y) {
```

Comment:
Is there a reason floorDiv and floorMod are public?

Reply:
Mostly for convenience. I've moved them to `MathUtil` in wpimath.